### ISSUE TITLE
inds variable referenced before assignment

### ISSUE THREAD

**Issue by aaraney** [kvanwerkhoven/rti-nwc#17]
@kvanwerkhoven can you have a look at this and tell me what this expected behavior should be here? If the `else` clause is hit in the first cycle of the loop you will get a referenced before assignment error.

https://github.com/kvanwerkhoven/rof/blob/1f3ed8b6a760b64c188bf3586b9f79af818d4de7/ROF/nwm_utilities.py#L1039

**Comment by aaraney** [comment]
Nevermind, for whatever reason the logs I was getting was not showing that the file read was not successful. I think it makes the most sense to instantiate the `inds` variable to a ~sentinel value like `None` before the loop and check for that afterwards to then raise an exception.~ it makes more sense to set it to False.

**Comment by aaraney** [comment]
So, it looks like the `analysis_assim_extend` [didn't make it to google cloud](https://console.cloud.google.com/storage/browser/national-water-model/nwm.20210525;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false) on 5/25/21 at 21:30Z.

**Comment by jarq6c** [comment]
> So, it looks like the `analysis_assim_extend` [didn't make it to google cloud](https://console.cloud.google.com/storage/browser/national-water-model/nwm.20210525;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false) on 5/25/21 at 21:30Z.

Looks like data is also missing internally. Could be having issues today.

**Comment by kvanwerkhoven** [comment]
Catching up.  If files do not exist, `success` should return false and skip that block.  So seems there is a bug within `get_flow_from_path`

**Comment by kvanwerkhoven** [comment]
trying to replicate issue on my local branch

**Comment by kvanwerkhoven** [comment]
So seems this could be a good example to switch to raising an exception.  Code appears to be working as intended, but as intended was not good design.  As it stands - it reads each file in the filelist, if file is not found, `success` is set to False but does not exit at that point. I use the sys.exit once it comes out of function `build_flow_array` if `success` if False (if any file not read successfully).  This all works on my end, but cannot remember why not just exit (raise exception) as soon as a single file was not found/read.  I think there was a reason, but cannot recall.

**Comment by kvanwerkhoven** [comment]
@aaraney Hey Austin - where did you add your comment about alternate method to raise exception?  I can't find it

**Comment by aaraney** [comment]
Hey Katie, here [it]( https://github.com/kvanwerkhoven/rof/issues/3#issuecomment-840734150 ) is. It took me a second to find them as well!

**Comment by kvanwerkhoven** [comment]
Wow, I read through that comment section twice and missed it!  See it now, thanks!

**Comment by aaraney** [comment]
Haha no worries @kvanwerkhoven!

**Comment by kvanwerkhoven** [comment]
@aaraney I tried the steps you laid out in PR #13 and got a conflict error (at bottom) - any suggestions?

(nwm) PS C:\repos\git\rof> git status
On branch rof_v1
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   ROF/nwm_utilities.py

no changes added to commit (use "git add" and/or "git commit -a")
(nwm) PS C:\repos\git\rof> git stash
Saved working directory and index state WIP on rof_v1: 0c3f9e1 config file upate
(nwm) PS C:\repos\git\rof> git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
(nwm) PS C:\repos\git\rof> git fetch
remote: Enumerating objects: 40, done.
remote: Counting objects: 100% (40/40), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 34 (delta 13), reused 19 (delta 7), pack-reused 0
Unpacking objects: 100% (34/34), 10.43 KiB | 77.00 KiB/s, done.
From https://github.com/kvanwerkhoven/rof
   b1a6142..1f3ed8b  main       -> origin/main
(nwm) PS C:\repos\git\rof> git rebase
Successfully rebased and updated refs/heads/main.
(nwm) PS C:\repos\git\rof> git checkout -b map-raise-updates
Switched to a new branch 'map-raise-updates'
(nwm) PS C:\repos\git\rof> git stash pop
Auto-merging ROF/nwm_utilities.py
CONFLICT (content): Merge conflict in ROF/nwm_utilities.py
The stash entry is kept in case you need it again.
(nwm) PS C:\repos\git\rof>

**Comment by aaraney** [comment]
Ah the dreaded merge conflict. So what that means in there are lines in `ROF/nwm_utilities.py` that differ in the stash vs the one in the `map-raise-updates`. Here are some helpful [docs](https://docs.github.com/en/github/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line#competing-line-change-merge-conflicts).

**Comment by aaraney** [comment]
5,6, and 7 are really what you need to pay attention too.

**Comment by aaraney** [comment]
Congrats on the first merge conflict!

**Comment by kvanwerkhoven** [comment]
OK - I kept both changes - one was my new 'figure_colors' function, the other the make_dir... function you added yesterday.  I don't quite understand why that was a 'conflict', but we can talk through sometime.  

I also added a new function "check_files" that is called after `download_from_google` completes for the full filelist.  Doing it there (after attempting all downloads) rather than killing it as soon as a missing file is encountered so that any files that exist (i.e., after the missing one in the list) still are downloaded and are available for later or for the other version (LatestAnA versus ExtAnA).

This should prevent the above bug (which I did eventually reproduce) since if any channel-rt files are missing, it will exit before calling `get_flow_array`.  But to be safe, added a check if `inds` is False (and set to False above) to prevent that bug directly.  So I think this issue can be closed.

**Comment by aaraney** [comment]
> OK - I kept both changes - one was my new 'figure_colors' function, the other the make_dir... function you added yesterday. I don't quite understand why that was a 'conflict', but we can talk through sometime.

Yeah we can cover that in our next meeting. Glad you got it figured out though.

> I also added a new function "check_files" that is called after download_from_google completes for the full filelist. Doing it there (after attempting all downloads) rather than killing it as soon as a missing file is encountered so that any files that exist (i.e., after the missing one in the list) still are downloaded and are available for later or for the other version (LatestAnA versus ExtAnA).

Awesome, I appreciate not taking a fail fast approach in this context, there is just so much data that needs to be downloaded!
